File: android/feature/nova/src/main/kotlin/com/riskdetectedan/feature/nova/NovaHomeDeadlineBoard.kt
package com.riskdetectedan.feature.nova

import androidx.compose.animation.animateContentSize
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.unit.dp
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.compose.LocalLifecycleOwner
import androidx.lifecycle.compose.currentStateAsState
import com.riskdetectedan.core.data.nova.NovaFollowupPage
import com.riskdetectedan.core.data.nova.NovaEquipmentWords
import com.riskdetectedan.core.designsystem.isg.*
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch

private enum class DeadlineBucket(val title: String, val symbol: String, val status: String, val empty: String,
                                  val background: NovaColorToken, val ink: NovaColorToken, val tag: String) {
    overdue("Süresi Geçenler", "exclamationmark.circle", "expired", "Süresi geçen iş yok",
        NovaColorToken.statusDangerBg, NovaColorToken.statusDangerInk, "overdue"),
    upcoming("Yaklaşan", "calendar.badge.clock", "soon", "Yaklaşan iş yok",
        NovaColorToken.statusInfoBg, NovaColorToken.statusInfoInk, "upcoming"),
}

private class DeadlineColumn {
    var rows by mutableStateOf<List<NovaFollowupPage.Row>>(emptyList())
    var count by mutableIntStateOf(0)
    var hasMore by mutableStateOf(false)
    var expanded by mutableStateOf(false)
}

/**
 * Record-level deadlines on the home page (iOS `NovaHomeDeadlineBoard`). The source is the same scoped
 * timeline Evrak Takibi reads; counts are never turned into fake cards. [load] asks for one status
 * ("expired" or "soon") from an offset; [onOpen] shows the record in place of the home page.
 */
@Composable
fun NovaHomeDeadlineBoard(load: suspend (status: String, offset: Int) -> NovaFollowupPage, changes: Flow<Unit>,
                          scopeKey: Any?, onPendingActionCount: (Int?) -> Unit = {}, onOpen: (NovaFollowupPage.Row) -> Unit) {
    val columns = remember { DeadlineBucket.entries.associateWith { DeadlineColumn() } }
    var loading by remember { mutableStateOf(true) }
    var failed by remember { mutableStateOf(false) }
    var loadingMore by remember { mutableStateOf<DeadlineBucket?>(null) }
    var revision by remember { mutableIntStateOf(0) }
    val coroutines = rememberCoroutineScope()

    // iOS refreshes when the scene turns active again and when a record changes.
    val lifecycle by LocalLifecycleOwner.current.lifecycle.currentStateAsState()
    var wasResumed by remember { mutableStateOf(true) }
    LaunchedEffect(lifecycle) {
        val resumed = lifecycle.isAtLeast(Lifecycle.State.RESUMED)
        if (resumed && !wasResumed) revision++
        wasResumed = resumed
    }
    LaunchedEffect(changes) { changes.collect { revision++ } }

    LaunchedEffect(scopeKey, revision) {
        loading = true; failed = false
        try {
            coroutineScope {
                val late = async { load(DeadlineBucket.overdue.status, 0) }
                val soon = async { load(DeadlineBucket.upcoming.status, 0) }
                val (latePage, soonPage) = late.await() to soon.await()
                columns.getValue(DeadlineBucket.overdue).apply { rows = latePage.rows; count = latePage.expired; hasMore = latePage.hasMore }
                columns.getValue(DeadlineBucket.upcoming).apply { rows = soonPage.rows; count = soonPage.soon; hasMore = soonPage.hasMore }
                onPendingActionCount(latePage.expired + soonPage.soon)
            }
        } catch (error: kotlinx.coroutines.CancellationException) {
            throw error
        } catch (_: Exception) {
            failed = true
            onPendingActionCount(null)
        }
        loading = false
    }

    fun loadMore(bucket: DeadlineBucket) {
        if (loadingMore != null) return
        val column = columns.getValue(bucket)
        loadingMore = bucket
        coroutines.launch {
            // Keep the already visible records on failure; the next tap can retry.
            runCatching { load(bucket.status, column.rows.size) }.getOrNull()?.let { page ->
                column.rows = column.rows + page.rows; column.hasMore = page.hasMore
            }
            loadingMore = null
        }
    }

    val content: @Composable (DeadlineBucket, Modifier) -> Unit = { bucket, modifier ->
        DeadlineColumnView(bucket, columns.getValue(bucket), loading, failed, loadingMore == bucket, modifier,
            onRetry = { revision++ }, onMore = { loadMore(bucket) }, onOpen = onOpen)
    }
    Box(Modifier.testTag("nova.home.deadlines")) {
        if (novaFontScaleIsAccessibility()) {
            Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
                DeadlineBucket.entries.forEach { content(it, Modifier.fillMaxWidth()) }
            }
        } else {
            Row(horizontalArrangement = Arrangement.spacedBy(10.dp), verticalAlignment = Alignment.Top) {
                DeadlineBucket.entries.forEach { content(it, Modifier.weight(1f)) }
            }
        }
    }
}

@Composable
private fun DeadlineColumnView(bucket: DeadlineBucket, column: DeadlineColumn, loading: Boolean, failed: Boolean, loadingMore: Boolean,
                               modifier: Modifier, onRetry: () -> Unit, onMore: () -> Unit, onOpen: (NovaFollowupPage.Row) -> Unit) {
    val ink = bucket.ink.color()
    val muted = NovaColorToken.textSecondary.color()
    val visible = if (column.expanded) column.rows else column.rows.take(3)
    Column(modifier.background(bucket.background.color(), RoundedCornerShape(20.dp)).padding(10.dp)
        .animateContentSize(tween(200)).testTag("nova.home.deadlines.${bucket.tag}"),
        verticalArrangement = Arrangement.spacedBy(10.dp)) {
        Row(Modifier.heightIn(min = 30.dp), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(5.dp)) {
            NovaIcon(bucket.symbol, 13.dp, tint = ink)
            NovaText(bucket.title, Modifier.weight(1f), NovaTypeToken.bodyStrong, ink, maxLines = 1)
            Box(Modifier.defaultMinSize(26.dp, 26.dp).background(ink.copy(alpha = 0.10f), CircleShape), contentAlignment = Alignment.Center) {
                NovaText(if (loading) "·" else column.count.toString(), style = NovaTypeToken.meta, color = ink)
            }
        }
        when {
            loading -> Box(Modifier.fillMaxWidth().heightIn(min = 66.dp), contentAlignment = Alignment.Center) {
                CircularProgressIndicator(Modifier.size(22.dp), color = ink, strokeWidth = 2.dp)
            }
            failed -> {
                NovaText("Kayıtlar yüklenemedi", Modifier.fillMaxWidth().heightIn(min = 66.dp), NovaTypeToken.meta, muted)
                Box(Modifier.heightIn(min = 36.dp).novaRowPress(onClick = onRetry), contentAlignment = Alignment.CenterStart) {
                    NovaText("Tekrar dene", style = NovaTypeToken.meta, color = ink)
                }
            }
            column.rows.isEmpty() -> NovaText(bucket.empty, Modifier.fillMaxWidth().heightIn(min = 66.dp), NovaTypeToken.meta, muted)
            else -> {
                visible.forEach { row -> DeadlineCard(row, ink) { onOpen(row) } }
                if (column.rows.size > 3 || column.hasMore) {
                    if (column.expanded && column.hasMore) {
                        Box(Modifier.fillMaxWidth().heightIn(min = 36.dp).novaRowPress(enabled = !loadingMore, onClick = onMore),
                            contentAlignment = Alignment.Center) {
                            NovaText(if (loadingMore) "Yükleniyor…" else "Daha fazla", style = NovaTypeToken.meta, color = ink)
                        }
                    }
                    Row(Modifier.fillMaxWidth().heightIn(min = 38.dp).novaRowPress { column.expanded = !column.expanded },
                        horizontalArrangement = Arrangement.spacedBy(4.dp, Alignment.CenterHorizontally), verticalAlignment = Alignment.CenterVertically) {
                        NovaText(if (column.expanded) "Daha az" else "Tümünü gör", style = NovaTypeToken.meta, color = ink)
                        NovaIcon(if (column.expanded) "chevron.up" else "chevron.right", 11.dp, tint = ink)
                    }
                }
            }
        }
    }
}

@Composable
private fun DeadlineCard(row: NovaFollowupPage.Row, ink: Color, onInspect: () -> Unit) {
    val equipmentName = row.equipmentTypeLabel?.trim()?.takeIf { it.isNotEmpty() }
        ?: row.equipmentType?.let(NovaEquipmentWords::type) ?: "Ekipman"
    val detail = if (row.kind == "equipment") "$equipmentName · ${row.title}" else row.title
    val shape = RoundedCornerShape(13.dp)
    Row(Modifier.fillMaxWidth().shadow(3.dp, shape, ambientColor = Color.Black.copy(alpha = 0.04f), spotColor = Color.Black.copy(alpha = 0.04f))
        .background(NovaColorToken.surface.color(), shape)
        .novaRowPress(onClick = onInspect).testTag("nova.home.deadline.inspect.${row.key}").padding(8.dp),
        verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(5.dp)) {
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(3.dp)) {
            NovaText(row.typeTitle, style = NovaTypeToken.badge, color = ink, maxLines = 1)
            NovaText(detail, style = NovaTypeToken.bodyStrong, color = NovaColorToken.text.color(), maxLines = 2)
            NovaText(row.companyName, style = NovaTypeToken.metaQuiet, color = NovaColorToken.textSecondary.color(), maxLines = 1)
            row.dueOn?.let { day ->
                NovaText(NovaDay.label(day), style = NovaTypeToken.badge, color = ink, maxLines = 1)
            }
        }
        Box(Modifier.size(26.dp).background(ink.copy(alpha = 0.10f), CircleShape), contentAlignment = Alignment.Center) {
            NovaIcon("chevron.right", 10.dp, tint = ink)
        }
    }
}

File: App/DesignSystem/ISG/NovaHomeDeadlineBoard.swift
import SwiftUI

/// Record-level deadlines on the home page. The source is the same scoped
/// timeline used by Evrak Takibi; counts are never turned into fake cards.
struct NovaHomeDeadlineBoard: View {
    let identity: NovaSessionIdentity
    let canWrite: Bool
    let scopeID: UUID?

    @Environment(\.colorScheme) private var scheme
    @Environment(\.dynamicTypeSize) private var typeSize
    @Environment(\.scenePhase) private var scenePhase
    @State private var overdue: [NovaFollowupPage.Row] = []
    @State private var upcoming: [NovaFollowupPage.Row] = []
    @State private var overdueCount = 0
    @State private var upcomingCount = 0
    @State private var overdueHasMore = false
    @State private var upcomingHasMore = false
    @State private var expandedOverdue = false
    @State private var expandedUpcoming = false
    @State private var loading = true
    @State private var loadingMore: String?
    @State private var failed = false
    @State private var revision = 0
    @State private var selected: NovaFollowupPage.Row?
    @State private var selectedTraining: NovaFollowupPage.Row?

    private var requestKey: String {
        "\(identity.userID):\(identity.sessionID):\(scopeID?.uuidString ?? "personal"):\(revision)"
    }

    var body: some View {
        Group {
            if typeSize.isAccessibilitySize {
                VStack(alignment: .leading, spacing: 12) {
                    column(.overdue)
                    column(.upcoming)
                }
            } else {
                HStack(alignment: .top, spacing: 10) {
                    column(.overdue)
                    column(.upcoming)
                }
            }
        }
        .task(id: requestKey) { await refresh() }
        .onReceive(NotificationCenter.default.publisher(for: Notification.Name("isgada.records.changed"))) { event in
            guard event.object as? UUID == identity.userID else { return }
            revision += 1
        }
        .onChange(of: scenePhase) { phase in
            if phase == .active { revision += 1 }
        }
        .novaPopup(item: $selected, onDismiss: { revision += 1 }) { row in
            NovaFollowupDestination(identity: identity, row: row, canWrite: canWrite,
                onBack: { selected = nil })
        }
        .novaFullScreenCover(item: $selectedTraining, onDismiss: { revision += 1 }) { row in
            NovaFollowupDestination(identity: identity, row: row, canWrite: canWrite,
                onBack: { selectedTraining = nil })
        }
        .accessibilityIdentifier("nova.home.deadlines")
    }

    private enum Bucket {
        case overdue, upcoming

        var title: String { self == .overdue ? "Süresi Geçenler" : "Yaklaşan" }
        var symbol: String { self == .overdue ? "exclamationmark.circle" : "calendar.badge.clock" }
        var status: String { self == .overdue ? "expired" : "soon" }
        var empty: String { self == .overdue ? "Süresi geçen iş yok" : "Yaklaşan iş yok" }
        var background: NovaColorToken { self == .overdue ? .statusDangerBg : .statusInfoBg }
        var ink: NovaColorToken { self == .overdue ? .statusDangerInk : .statusInfoInk }
    }

    private func column(_ bucket: Bucket) -> some View {
        let rows = bucket == .overdue ? overdue : upcoming
        let count = bucket == .overdue ? overdueCount : upcomingCount
        let expanded = bucket == .overdue ? expandedOverdue : expandedUpcoming
        let hasMore = bucket == .overdue ? overdueHasMore : upcomingHasMore
        let visible = expanded ? rows : Array(rows.prefix(3))
        let ink = bucket.ink.color(in: scheme)

        return VStack(alignment: .leading, spacing: 10) {
            HStack(spacing: 5) {
                Image(systemName: bucket.symbol)
                    .font(.system(size: 13, weight: .semibold))
                Text(bucket.title)
                    .font(NovaFont.font(.bodyStrong))
                    .lineLimit(1)
                    .minimumScaleFactor(0.75)
                Spacer(minLength: 2)
                Text(loading ? "·" : String(count))
                    .font(NovaFont.font(.meta))
                    .frame(minWidth: 26, minHeight: 26)
                    .background(ink.opacity(0.10), in: Circle())
            }
            .foregroundStyle(ink)
            .frame(minHeight: 30)

            if loading {
                ProgressView()
                    .frame(maxWidth: .infinity, minHeight: 66)
            } else if failed {
                Text("Kayıtlar yüklenemedi")
                    .font(NovaFont.font(.meta))
                    .foregroundStyle(NovaColorToken.textSecondary.color(in: scheme))
                    .frame(maxWidth: .infinity, minHeight: 66, alignment: .leading)
                Button("Tekrar dene") { revision += 1 }
                    .font(NovaFont.font(.meta))
                    .foregroundStyle(ink)
            } else if rows.isEmpty {
                Text(bucket.empty)
                    .font(NovaFont.font(.meta))
                    .foregroundStyle(NovaColorToken.textSecondary.color(in: scheme))
                    .frame(maxWidth: .infinity, minHeight: 66, alignment: .leading)
            } else {
                ForEach(visible) { row in
                    deadlineCard(row, ink: ink)
                }
                if rows.count > 3 || hasMore {
                    if expanded && hasMore {
                        Button {
                            Task { await loadMore(bucket) }
                        } label: {
                            Text(loadingMore == bucket.status ? "Yükleniyor…" : "Daha fazla")
                                .font(NovaFont.font(.meta))
                                .foregroundStyle(ink)
                                .frame(maxWidth: .infinity, minHeight: 36)
                        }
                        .buttonStyle(.plain)
                        .disabled(loadingMore != nil)
                    }
                    Button {
                        withAnimation(.easeInOut(duration: 0.2)) { setExpanded(bucket, !expanded) }
                    } label: {
                        HStack(spacing: 4) {
                            Text(expanded ? "Daha az" : "Tümünü gör")
                            Image(systemName: expanded ? "chevron.up" : "chevron.right")
                        }
                        .font(NovaFont.font(.meta))
                        .foregroundStyle(ink)
                        .frame(maxWidth: .infinity, minHeight: 38)
                    }
                    .buttonStyle(.plain)
                }
            }
        }
        .padding(10)
        .frame(maxWidth: .infinity, alignment: .topLeading)
        .background(bucket.background.color(in: scheme), in: RoundedRectangle(cornerRadius: 20))
        .accessibilityIdentifier(bucket == .overdue ? "nova.home.deadlines.overdue" : "nova.home.deadlines.upcoming")
    }

    private func deadlineCard(_ row: NovaFollowupPage.Row, ink: Color) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            Text(row.title)
                .font(NovaFont.font(.bodyStrong))
                .foregroundStyle(NovaColorToken.text.color(in: scheme))
                .lineLimit(2)
                .fixedSize(horizontal: false, vertical: true)
            Text(row.company_name)
                .font(NovaFont.font(.meta))
                .foregroundStyle(NovaColorToken.textSecondary.color(in: scheme))
                .lineLimit(2)
            if let date = row.due_on {
                Label(NovaStatisticsSnapshot.dayLabel(date), systemImage: "calendar")
                    .font(NovaFont.font(.meta))
                    .foregroundStyle(ink)
                    .lineLimit(1)
                    .minimumScaleFactor(0.8)
            }
            Button {
                if row.kind == "training" { selectedTraining = row }
                else { selected = row }
            } label: {
                HStack(spacing: 4) {
                    Text("İncele")
                    Image(systemName: "chevron.right")
                        .font(.system(size: 10, weight: .bold))
                }
                .font(NovaFont.font(.meta))
                .foregroundStyle(.white)
                .frame(maxWidth: .infinity, minHeight: 36)
                .background(ink, in: Capsule())
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("nova.home.deadline.inspect.\(row.id)")
        }
        .padding(10)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(NovaColorToken.surface.color(in: scheme), in: RoundedRectangle(cornerRadius: 15))
        .shadow(color: .black.opacity(0.04), radius: 6, y: 2)
    }

    private func setExpanded(_ bucket: Bucket, _ value: Bool) {
        if bucket == .overdue { expandedOverdue = value }
        else { expandedUpcoming = value }
    }

    @MainActor private func refresh() async {
        loading = true
        failed = false
        let service = NovaFollowupService(identity: identity)
        do {
            async let overduePage = service.load(company: nil, status: "expired")
            async let upcomingPage = service.load(company: nil, status: "soon")
            let (late, soon) = try await (overduePage, upcomingPage)
            try Task.checkCancellation()
            overdue = late.rows
            upcoming = soon.rows
            overdueCount = late.expired
            upcomingCount = soon.soon
            overdueHasMore = late.has_more
            upcomingHasMore = soon.has_more
        } catch {
            if !Task.isCancelled { failed = true }
        }
        if !Task.isCancelled { loading = false }
    }

    @MainActor private func loadMore(_ bucket: Bucket) async {
        guard loadingMore == nil else { return }
        loadingMore = bucket.status
        defer { loadingMore = nil }
        do {
            let offset = bucket == .overdue ? overdue.count : upcoming.count
            let page = try await NovaFollowupService(identity: identity).load(
                company: nil, status: bucket.status, offset: offset)
            try Task.checkCancellation()
            if bucket == .overdue {
                overdue += page.rows
                overdueHasMore = page.has_more
            } else {
                upcoming += page.rows
                upcomingHasMore = page.has_more
            }
        } catch {
            // Keep the already visible records; the next tap can retry.
        }
    }
}
